_posts/2020-11-09-defensive-programing.md

```diff
@@ -0,0 +1,110 @@
+---
+layout: post
+title:  "Defensive programming - what it is and why it is important"
+date:   2020-11-09 12:54:04 -0700
+categories: defensive programming, assertions, pre-conditions, nulls
+---
+# **Defensive programming - what it is and why it is important**
+Gabrielly Lunkes  
+
+![Dwight Schrute from The Office show saying "HE PROTEC"]({{site.url }}https://philippegroarke.com/img/2018/defensive.png)
+
+Often times, a piece of code is used by different people, not only the person who created it. Also, codes can frequently be used to process different types of data. In my opinion, this flexibility to use the same code to process different datasets is one of the greatest things about coding. However, how is it possible to write a complex code in a way that other users would be able to easily navigate through and even use it with different types of input without bugs? Yes, you guessed it, the answer is in defensive programming.
+
+## What is defensive programming?
+
+In short, defensive programming is a way to make sure that your code works even when given incorrect input. This can be done by adding guard statements and assertions in your code that check if certain requirements are met and return a statement. The statements are usually helpful sentences that identify which requirement was not met and can even tell you what you should do to fix it. Here is an example:
+```r
+if (nrow(input_data) == 0) {
+  stop("ERROR: Input data is empty. Did you make a typo?")
```

Comment:
You're missing the closing curly brace in this code example.

Reply:
Ok, I added it!